lib/Cro/APIToken/Store/Pg.pm6

```diff
@@ -24,27 +42,28 @@ class Cro::APIToken::Store::Pg does Cro::APIToken::Store {
     }
 
     method find-tokens(Cro::APIToken::Manager $manager, :%metadata,
-                       Bool :$expired = False, Bool :$revoked --> Seq) {
-        return ().Seq unless self!check-db-table-presence();
-        my $ordered-values = %metadata.keys.cache;
-        my $metadata-format = $ordered-values.kv.map(-> $k, $v { "(metadata->>'$v') = \${ $k + 1 }" }).join(' AND ');
+                       Bool :$expired = False, Bool :$revoked = True --> Seq) {
+        my $ordered-keys = %metadata.keys.cache;
+        for @$ordered-keys -> $key {
+            die X::Cro::APIToken::Store::Pg::DangerousKey.new(:$key) unless $key ~~ /<:L+[_-]>/;
+        }
+        my $metadata-format = $ordered-keys.kv.map(-> $k, $v { "(metadata->>'$v') = \${ $k + 1 }" }).join(' AND ');
         $metadata-format = 'AND ' ~ $metadata-format if $metadata-format;
 
-        my $query-text = "SELECT * FROM $!table-name WHERE { $expired ?? 'true' !! 'expiration > now()::date' } { $revoked ?? 'AND true' !! 'AND revoked = FALSE' } " ~
+        my $query-text = "SELECT * FROM $!table-name WHERE { $expired ?? 'TRUE' !! 'expiration IS NULL OR expiration > now()::date' } { $revoked ?? '' !! 'AND revoked = FALSE' } " ~
                 $metadata-format ~ ';';
-        my $rows = $!handle.query($query-text, |$ordered-values.map({ %metadata{$_} })).hashes;
+        my $rows = $!handle.query($query-text, |$ordered-keys.map({ %metadata{$_} })).hashes;
         $rows.map({ Cro::APIToken::Token.new(:$manager, |$_) });
     }
 
-    method revoke(Str $token --> Nil) {
-        return unless self!check-db-table-presence();
-        $!handle.query("UPDATE $!table-name SET revoked = TRUE WHERE token = \$1;", $token);
+    method revoke-token(Cro::APIToken::Token $token --> Nil) {
+        $!handle.query("UPDATE $!table-name SET revoked = TRUE WHERE token = \$1;", $token.token);
     }
 
     method !check-db-table-presence() {
-        my $table-exists = $!handle.query("SELECT EXISTS (SELECT FROM information_schema.tables WHERE table_name = 'croapitokenstorepg');").value;
+        my $table-exists = $!handle.query("SELECT EXISTS (SELECT FROM information_schema.tables WHERE table_name = '$!table-name');").value;
         if !$table-exists && $!create-table {
-            $!handle.execute('CREATE table CroAPITokenStorePg (id serial NOT NULL PRIMARY KEY, token text NOT NULL, expiration timestamp NOT NULL, metadata json, revoked boolean);');
+            $!handle.execute("CREATE table $!table-name (id serial NOT NULL PRIMARY KEY, token text NOT NULL, expiration timestamp, metadata jsonb, revoked boolean);");
```

Comment:
By using `CREATE TABLE IF NOT EXISTS $!table-name ... ` you could get rid of the `SELECT` call above.

Reply:
Yes, but it will break the logic "If ordered to create, create, but otherwise do not create and just complain".
Yes, we can add the "IF NOT EXISTS" bit conditionally and handle the flow in some other way, but it will become more confusing instead of the straightforward way done here.
Given the token storage is something created at init of the application and these two calls are executed during object construction, making the code harder to read for such a slow path is maybe not the best idea (though can be applied if somehow hot in future).

Reply:
The goal is to:

* Always check the table exists at construction time, so we get errors sooner than the first query (probably at startup time)
* Optionally create the table automatically, if `create-table` is `True`

The latter may not be desirable if one is using something else to manage to database schema as a whole. The vast majority of the time it's only going to be the "does the table exist" part that runs, which is presumably easier to parse than the whole create table query, although it's all very marginal given this happens once at application startup.